Dynamo+LTC: merging related code from staging branch to master

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #75046

Merge the code needed for dynamic+ltc integration from the staging branch to the master branch.


Test plan:
Unit test
```
pytest test_extract_compiled_graph
```
test thru dynamo
```
LTC_TS_CUDA=1 time python torchbench.py --speedup-ltc -dcuda --nvfuser --randomize-input --only <model name>
```

release notes: lazy
topics: not user facing

Differential Revision: [D35300646](https://our.internmc.facebook.com/intern/diff/D35300646)

shunting314 committed:
Dynamo+LTC: merging related code from staging branch to master

[ghstack-poisoned]

shunting314 commented:
Previous comments is in https://github.com/pytorch/pytorch/pull/75044

wconstab reviewed (approved):
LGTM pending some fixes commented in the previous PR.  Also pending if @Krovatkin wants to review.

shunting314 commented:
@shunting314 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35300646).

Krovatkin commented on torch/csrc/lazy/python/init.cpp:
```diff
@@ -17,6 +20,12 @@
 namespace torch {
 namespace lazy {
 
+struct NoGilSection {
```
we should be using the pybind11 equivalent : `py::gil_scoped_release `

Krovatkin commented on torch/_lazy/computation.py:
```diff
@@ -0,0 +1,15 @@
+import torch._C._lazy
```
it seems a bit arbitrary that these are grouped together under `computation`.  For example, why didn't `sync_multi` make the cut into this file?

shunting314 replied:
yes, I don't find a hard rule to decide which APIs goes to which module. This file specifically contains APIs needed to reused a compiled graph so far.  `sync_multi` is a more general API thus I didn't add that here. But it should also look good to put `sync_multi` here

Krovatkin reviewed (approved):
a few nits

shunting314 committed:
Update on "Dynamo+LTC: merging related code from staging branch to master"


Merge the code needed for dynamic+ltc integration from the staging branch to the master branch.


Test plan:
Unit test
```
pytest test_extract_compiled_graph
```
test thru dynamo
```
LTC_TS_CUDA=1 time python torchbench.py --speedup-ltc -dcuda --nvfuser --randomize-input --only <model name>
```

Differential Revision: [D35300646](https://our.internmc.facebook.com/intern/diff/D35300646)

[ghstack-poisoned]

shunting314 commented on torch/_lazy/debug.py:
```diff
@@ -0,0 +1,20 @@
+       The text can be processed by tools like dot to be rendered in pdf,png etc."""
+    return torch._C._lazy._get_tensors_dot(tensors)
+
+def dump_ir(tensors, ir_format):
```
@wconstab  I've made the requested changes in this pr.

shunting314 committed:
Update on "Dynamo+LTC: merging related code from staging branch to master"


Merge the code needed for dynamic+ltc integration from the staging branch to the master branch.


Test plan:
Unit test
```
pytest test_extract_compiled_graph
```
test thru dynamo
```
LTC_TS_CUDA=1 time python torchbench.py --speedup-ltc -dcuda --nvfuser --randomize-input --only <model name>
```

Differential Revision: [D35300646](https://our.internmc.facebook.com/intern/diff/D35300646)

[ghstack-poisoned]

shunting314 commented:
@shunting314 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35300646).

shunting314 committed:
Update on "Dynamo+LTC: merging related code from staging branch to master"


Merge the code needed for dynamic+ltc integration from the staging branch to the master branch.


Test plan:
Unit test
```
pytest test_extract_compiled_graph
```
test thru dynamo
```
LTC_TS_CUDA=1 time python torchbench.py --speedup-ltc -dcuda --nvfuser --randomize-input --only <model name>
```

Differential Revision: [D35300646](https://our.internmc.facebook.com/intern/diff/D35300646)

[ghstack-poisoned]

shunting314 commented:
@shunting314 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35300646).

shunting314 committed:
Update on "Dynamo+LTC: merging related code from staging branch to master"


Merge the code needed for dynamic+ltc integration from the staging branch to the master branch.


Test plan:
Unit test
```
pytest test_extract_compiled_graph
```
test thru dynamo
```
LTC_TS_CUDA=1 time python torchbench.py --speedup-ltc -dcuda --nvfuser --randomize-input --only <model name>
```

Differential Revision: [D35300646](https://our.internmc.facebook.com/intern/diff/D35300646)

[ghstack-poisoned]

shunting314 committed:
Update on "Dynamo+LTC: merging related code from staging branch to master"


Merge the code needed for dynamic+ltc integration from the staging branch to the master branch.


Test plan:
Unit test
```
pytest test_extract_compiled_graph
```
test thru dynamo
```
LTC_TS_CUDA=1 time python torchbench.py --speedup-ltc -dcuda --nvfuser --randomize-input --only <model name>
```

Differential Revision: [D35300646](https://our.internmc.facebook.com/intern/diff/D35300646)

[ghstack-poisoned]

shunting314 commented:
@shunting314 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35300646).
